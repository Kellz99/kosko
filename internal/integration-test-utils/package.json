{
  "name": "@kosko/integration-test-utils",
  "version": "0.0.0",
  "private": true,
  "engines": {
    "node": ">=14"
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "exports": {
    "import": "./dist/index.mjs",
    "require": "./dist/index.js"
  },
  "dependencies": {
    "execa": "^5.1.1",
    "tslib": "^2.4.0"
  }
}
