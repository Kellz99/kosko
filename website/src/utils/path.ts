export function isAbsolute(path: string): boolean {
  return path.startsWith("/");
}

export function isRelative(path: string): boolean {
  return path.startsWith("./") || path.startsWith("../");
}

export function dirname(path: string): string {
  const index = path.lastIndexOf("/");
  if (index === -1) return path;
  return path.substring(0, index - 1) || "/";
}

export function extname(path: string): string {
  const index = path.lastIndexOf(".");
  if (index === -1) return "";
  return path.substring(index);
}

export function basename(path: string, ext?: string): string {
  const index = path.lastIndexOf("/");
  const base = index === -1 ? path : path.substring(index + 1);

  if (base.endsWith(ext)) {
    return base.substring(0, base.length - ext.length);
  }

  return base;
}

export function relative(from: string, to: string): string {
  const fromParts = from.split("/").filter(Boolean);
  const toParts = to.split("/").filter(Boolean);

  while (fromParts[0] && toParts[0]) {
    fromParts.shift();
    toParts.shift();
  }

  while (toParts[0] === "." || toParts[0] === "..") {
    const toPart = toParts.shift();

    if (toPart === "..") {
      fromParts.pop();
    }
  }

  while (fromParts.pop()) {
    toParts.unshift("..");
  }

  return toParts.join("/");
}
