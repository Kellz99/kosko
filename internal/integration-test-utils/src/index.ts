import execa from "execa";
import { dirname, join } from "path";
import fs from "fs/promises";

const root = join(__dirname, "../../..");

export async function runCLI(
  args: string[],
  options: execa.Options = {}
): Promise<execa.ExecaReturnValue> {
  return execa(join(root, "packages", "kosko", "bin", "kosko.js"), args, {
    ...options,
    env: {
      LC_ALL: "en_US",
      ...options.env
    }
  });
}

export async function installPackage(
  path: string,
  name: string
): Promise<void> {
  const src = join(root, "packages", name);
  const dest = join(path, "node_modules", "@kosko", name);

  await fs.mkdir(dirname(dest), { recursive: true });

  try {
    await fs.unlink(dest);
  } catch (err: any) {
    if (err.code !== "ENOENT") throw err;
  }

  await fs.symlink(src, dest, "dir");
}
