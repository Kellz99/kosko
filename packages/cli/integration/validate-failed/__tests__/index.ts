/// <reference types="jest-extended"/>
import execa from "execa";
import { dirname } from "path";
import { runCLI, installPackage } from "@kosko/integration-test-utils";

const testDir = dirname(__dirname);

let result: execa.ExecaReturnValue;

beforeAll(async () => {
  await installPackage(testDir, "env");
});

beforeEach(async () => {
  result = await runCLI(["validate"], {
    cwd: testDir,
    reject: false
  });
});

test("should return status code 1", () => {
  expect(result.exitCode).toEqual(1);
});

test("should print the error", () => {
  expect(result.stderr).toContain(
    "ValidationError: data/spec must have required property 'containers'"
  );
});
