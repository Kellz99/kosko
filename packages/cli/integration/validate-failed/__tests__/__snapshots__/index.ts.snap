// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should print the error 1`] = `
"
[1mcomponents/foo.js[22m - [31mâœ– 1[39m

  Kind: v1/Pod, Name: foo
    [31mâœ–[39m /spec must have required property 'containers'

[31merror[39m - Generate failed (Total [31mâœ– 1[39m)"
`;
